Added 'Loading sequence best practices' (#252)
playcanvas/developer.playcanvas.com: content/en/user-manual/optimization/load-time.md

position: 5
---

> Optimizing isn't just related to improving frame rate. Fast load times are also critical. The faster your app loads, the more likely your users will stick around to experience it. For example, users on mobiles for browser based games will wait 4-5 secs of inactivity before leaving.

Reviewer: Could that claim link to a citation?

Author: The only links I have are for websites: https://wemakewebsites.com/blog/improve-page-load-speed-increase-conversion/

Or from Facebook: https://developers.facebook.com/docs/games/instant-games/guides/game-performance/

If we don't mind linking to Facebook, I can rephrase and make it fit?

Reviewer: Instead of making a claim, I would just make a recommendation. `Aim to have your app load in less than 5 seconds to prevent users from churning.`

Author: Yeah, that sounds good

> Optimizing isn't just related to improving frame rate. Fast load times are also critical. The faster your app loads, the more likely your users will stick around to experience it. For example, users on mobiles for browser based games will wait 4-5 secs of inactivity before leaving.

Author: ```suggestion
Optimizing isn't just related to improving frame rate. Fast load times are also critical. The faster your app loads, the more likely your users will stick around to experience it. Aim to have your app load in less than 5 seconds to prevent users from churning.
```

* Using PlayCanvas' built-in physics engine incurs an additional download cost of 379KB. If you are using the physics engine to solve very simple problems, consider rolling an alternative solution that doesn't incur the download penalty.
* If you self-host your PlayCanvas app, be sure to configure your web server to serve files with GZIP compression. In particular, JSON and JS files.

## Loading sequence best practices

Beyond the above guidelines, it's possible to retain users by spacing the loading in multiple stages while giving the users something new to interact with or watch.

Using [Virtual Voodoo from MiniClip][1] which was made with PlayCanvas, we can show the 'typical' sequence that most applications will use for browser experiences.

The game has 3 phases:

1. Preloader
2. Title Screen and Character Customization
3. Main Game

<img src="/images/user-manual/optimization/loading/virtual-voodoo-phases.jpg" style="max-width: 100%;" />

The Preloader phase loads the assets that are needed for the first PlayCanvas scene which is the Title Screen and Character Customization. This would include assets for the UI, character model and assets.

> When the Title Screen is active, the game starts it's second load of assets in the background that are needed for the Main Game. The idea is with the transitions for the Title Screen and possible interaction of the character customization, by the time the user presses the start button, the assets would have finished loading.

Reviewer: ```suggestion
When the Title Screen is active, the game starts background loading the assets that are needed for the Main Game. During the transition to the Title Screen and possible interaction with the character customization, by the time the user presses the start button, the assets for the Main Game may have already finished loading.
```

However, if the user presses the start button before the assets have finished loading, a progress bar will appear on the button instead. Once it reaches 100%, the game will automatically transition to the Main Game.

<img src="/images/user-manual/optimization/loading/virtual-voodoo-assets-not-ready.gif" style="max-width: 480px;" />

> With the assets being loaded in phases and giving something new for the user to interact and/or look at periodically, the user stays engaged despite a long loading time.

Reviewer: ```suggestion
With the assets being loaded in phases and giving something new for the user to interact with and/or look at periodically, the user stays engaged despite a long loading time.
```

### Further improvements

> Some developers will go as far as to reduce the reduce the Preloader phases to only load the bare minimum and add an 'in-application' loading screen that allows the developer to populate with application related assets and text, use animation, etc. This engages the user as they are seeing something that is directly related to the application.

Reviewer: ```suggestion
Some developers will go as far as to reduce the Preloader phase to only load the bare minimum and add an 'in-application' loading screen that allows the developer to populate with application related assets and text, use animation, etc. This engages the user as they are seeing something that is directly related to the application.
```
